protocol/messenger_test.go

```diff
@@ -2540,3 +2540,237 @@ func (s *MessengerSuite) TestSendMessageMention() {
 	s.Require().NoError(err)
 	s.Require().Equal("Alice talk to bobby", response.Notifications()[0].Message)
 }
+
+func (s *MessengerSuite) TestMarkMessageAsUnread() {
+
+	s.T().Run("Test marking a message as unread when message list contains 1 message", func(t *testing.T) {
+		chat := CreatePublicChat("test-chat-1", s.m.transport)
+
+		chat.UnviewedMessagesCount = 2
+		chat.UnviewedMentionsCount = 2
+		chat.Highlight = true
+
+		err := s.m.SaveChat(chat)
+		s.Require().NoError(err)
+
+		inputMessage1 := buildTestMessage(*chat)
+		inputMessage1.ID = "1"
+		inputMessage1.Seen = true
+		inputMessage1.Mentioned = true
+
+		err = s.m.SaveMessages([]*common.Message{inputMessage1})
+		s.Require().NoError(err)
+
+		_, err = s.m.MarkAllRead(context.Background(), chat.ID)
+		s.Require().NoError(err)
+
+		chats := s.m.Chats()
+
+		for idx := range chats {
+			if chats[idx].ID == chat.ID {
+				s.Require().Equal(uint(0), chats[idx].UnviewedMessagesCount)
+				s.Require().Equal(uint(0), chats[idx].UnviewedMentionsCount)
+				s.Require().Equal(false, chats[idx].Highlight)
+			}
+		}
+
+		count, countWithMentions, err := s.m.MarkMessageAsUnread(chat.ID, inputMessage1.ID)
+		s.Require().NoError(err)
+		s.Require().Equal(uint64(1), count)
+		s.Require().Equal(uint64(1), countWithMentions)
+
+		chats = s.m.Chats()
+
+		for idx := range chats {
+			if chats[idx].ID == chat.ID {
+				s.Require().Equal(uint(1), chats[idx].UnviewedMessagesCount)
+				s.Require().Equal(uint(1), chats[idx].UnviewedMentionsCount)
+				s.Require().Equal(false, chats[idx].Highlight)
+			}
+		}
```

Comment:
Here you check 2 things:
- return value of `MarkMessageAsUnread`
- chat's `UnviewedMessagesCount` and `UnviewedMentionsCount`

It's good, but I think the most important part is to check that the messages `Seen` value was indeed updated. And that it's correct for each message. You can use `MessageByID` function for this. I think you have IDs of all saved messages.